### PR TITLE
Profiling, major performance improvements, nodemon crash fix

### PR DESCRIPTION
**Pull request by FredericaBernkastel**
**commit b48f430b427c380f5be8f4cbc7e7aa78ee629a73:**
Simple and useful profilier template, will output to console certain timings, only if node_env set to "development".
Tell me if it gives a problem to you, i'll ditch out from pull request.

**commit fe4f29cd10d5b80673214999a244b129402b3af8:**
I think the problem was with left joins. I rewrote the query in different way, which leads to **~15.5x speedup**.
For my test dataset, old timings were [700ms/1300ms](/mserajnik/hydrusrv/issues/4#issuecomment-437576268), and new ones are 45ms/90ms.
Doesn't depens on query complexity: i wrote a query with 30 parameters in total, timings have not changed.
"getByExcludeTags" is a bit faster too: before 315ms, after - 210ms.

Consider testing on your datasets aswell!

**commit b011725382885f6e91cb54c86d30cbde8ec741d7:**
Fixed nodemon crash.

### CODE REVIEW

**Comment by mserajnik** [comment]
Thanks 👍

Query times sound great and I'll test it on my data as well, but as I've mentioned [in the issue](https://github.com/mserajnik/hydrusrv/issues/4#issuecomment-437631882), the `getByExcludeTags()` function should still return files that have no tags at all and don't drop them when excluding tags (as it would with these changes). Could you take a look at that?

Not sure about the profiling, certainly useful, but I'd rather put it in its own file (`helpers/profiler.js`, then export the object and require in `helpers/data.js`, then call via `profiler.init()`, `profiler.log()` etc.). I suggest you keep this as is and I make those changes in a separate commit after merging, is that ok for you?

The nodemon behaviour is weird, it states [in its readme](https://github.com/remy/nodemon#ignoring-files) that `node_modules` should already be excluded by default. But if that works for you by specifically ignoring it when running, they probably have some bug, so this is fine.

---

Aside from that, I'll want to make sure the style is consistent with the rest of the application, so if it's alright with you, I'd make those adjustments after merging, saves me the trouble of reviewing each section separately and you don't have to make changes.

File: server/models/files.js [hunk 1]
```diff
@@ -144,35 +136,32 @@ module.exports = {
...
+      where 
+        hydrusrv_files.tags_id in (
+          select file_tags_id from hydrusrv_mappings
+            except
+          select file_tags_id from (
+            select file_tags_id, tag_id from hydrusrv_mappings
+            where tag_id in (
+              select id from hydrusrv_tags 
+              where name in (${',?'.repeat(excludeTags.length).replace(',', '')})
+            )
```
Comment (mserajnik) [inline server/models/files.js]: Please make sure this doesn't exclude files that have no tags, as also stated [here](https://github.com/mserajnik/hydrusrv/pull/5#issuecomment-437633217).

**Comment by FredericaBernkastel** [comment]
> but I'd rather put it in its own file (helpers/profiler.js, then export the object and require in helpers/data.js, then call via profiler.init(), profiler.log() etc.)

Yes, i wasn't sure about proper way to do this.

> getByExcludeTags() function should still return files that have no tags at all

Fixed!

**Comment by mserajnik** [comment]
From testing on the demo server, looks good to me:

Before:

+ `GET /api/files?tags[0]=scenery&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 200ms
+ `GET /api/files?tags[0]=scenery&tags[1]=-tree&tags[2]=-house&tags[3]=-car&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 206ms
+ `GET /api/files?tags[0]=-scenery&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 160ms
+ `GET /api/files?tags[0]=tree&tags[1]=sky&tags[2]=forest&tags[3]=river&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 184ms
+ `GET /api/files?tags[0]=tree&tags[1]=sky&tags[2]=forest&tags[3]=river&tags[4]=-bridge&tags[5]=-mountain&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 187ms

After:

+ `GET /api/files?tags[0]=scenery&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 91ms
+ `GET /api/files?tags[0]=scenery&tags[1]=-tree&tags[2]=-house&tags[3]=-car&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 110ms
+ `GET /api/files?tags[0]=-scenery&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 85ms
+ `GET /api/files?tags[0]=tree&tags[1]=sky&tags[2]=forest&tags[3]=river&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 96ms
+ `GET /api/files?tags[0]=tree&tags[1]=sky&tags[2]=forest&tags[3]=river&tags[4]=-bridge&tags[5]=-mountain&sort=namespaces&namespaces[0]=creator&namespaces[1]=series&namespaces[2]=title&namespaces[3]=volume&namespaces[4]=character&namespaces[5]=page&page=1` – 93ms

This was over the internet, so you'll probably have to substract around 30-50ms or so for latency. Would test locally on a larger set as well to get more exact results if I had one at hand atm, but honestly, I think it's clear enough that it's a big improvement, especially on larger databases (going by your examples).

I will do some additional local testing on a small set, just to make sure everything still works as intended (the DB tests clearly need an overhaul and I won't trust them anymore for the time being, that's why I've opened #6).

If nothing comes up there, I'll merge this into `develop` in the coming days, move the profiler into its own file, do the reformatting and make a new release.

Good work, thanks!

**Comment by FredericaBernkastel** [comment]
I've got a cool idea how to speed up it even further!
Wait a bit.

**Comment by mserajnik** [comment]
Sounds great! I'll take a look in the coming days.

**Comment by mserajnik** [comment]
Fyi, the [demo](https://hydrusrvue.mser.at) is now running with these changes, looks fine so far. I hope I'll get around to review and merge by Friday or Saturday, sorry to keep you waiting.

**Comment by mserajnik** [comment]
Unexpectedly found some time today, so I tested and merged this.

Everything looked good when testing locally.

Update time went down from around 1100ms (for a small set of 2k files, 40k mappings) to 200ms. Not the huge increase you've seen, but that's probably just because it was already decently quick for smaller data sets before.

Query performance is pretty amazing, testing locally (so no network delay aside the usual HTTP overhead) I was getting 15-20ms (instead of around 60ms) on some queries.

All in all, great work, thank you very much! ♥️

Will do the reformatting now and make a new release later in the evening.

**Comment by FredericaBernkastel** [comment]
Is this guy with 200k files gonna test it aswell?
Really wondering, how faster will it be for his case.

**Comment by Robert99943** [comment]
No problem, and yes the performance is much better now. Update times are around ~80 seconds and file query times are about ~600ms for basic 1-2 tags no exclusion and ~1000ms with exclusion.

**Comment by FredericaBernkastel** [comment]
Interesting, can i have this database aswell, in order to do further optimizations?

**Comment by Robert99943** [comment]
Sure, here is my current db. Grew a bit, 216k files, 25m mappings.
https://koto.reisen/serverdb.7z

**Comment by FredericaBernkastel** [comment]
Detailed timings for the future references:
```
get namespaces: 22.413s
fill temporary namespaces table: 0.001s
fill temporary tags table: 18.838s
fill temporary files table: 21.260s
fill temporary mappings table: 32.971s
replace current temporary tables: 0.002s
total: 95.494s

[ namespaces: 15, files: 73240, tags: 216813, mappings: 7655592 ]
```

**Comment by mserajnik** [comment]
Hm, are we missing files here? Careful, iirc on Robert‘s installation, 2 is the files repository and 3 is the tags repository.

**Comment by mserajnik** [comment]
Ok, pretty sure it was the other way around in the older dump he gave me, but this might be a fresh database anyway.

Regardless, unless Robert swapped the repository types (which I‘m not sure is possible in the hydrus server administration) so new files go the former tags repo and vice versa and/or he has a huge amount of unsupported file types (probably not), I‘m curious where the other 143k files and their mappings went.

**Comment by FredericaBernkastel** [comment]
If i swap file and tag ids, it just crashes.

**Comment by Robert99943** [comment]
Yeah sorry, this is a fresh db.

**Pull request merged**